mautic/mautic
# Export to CSV and Excel creates contact duplicates

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Yes]
| New feature/enhancement? (use the a.x branch)      | [No]
| Deprecations?                          | [No]
| BC breaks? (use the c.x branch)        | [No]
| Automated tests included?              | [No] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
This PR solves the issue [Export to CSV and Excel creates contact duplicates](https://forum.mautic.org/t/export-to-csv-and-excel-creates-contact-duplicates/22647).

When there are more than 10,000 client records in the DB and one of the fields is null for all (most) records, selectining data doesn't work correctly. Some contacts are duplicated, some are not selected. As a result, the data isn't imported to CSV and Excel correctly.

#### Reproduce the issue:

1. Install the new instance
2. Import the mock data (10,000 or more). It can be the data below
![image](https://user-images.githubusercontent.com/96085911/153473508-f78d9329-4c60-4a15-ae93-8df820d1cf1c.png)

2. Open the contact list
3. Leave the default sorting or order by the field with no value (e.g. last_activity if the data was imported)
![image](https://user-images.githubusercontent.com/96085911/153473864-92390a36-4a32-4c79-84fb-dfcdd2578bc0.png)

4. Export to CVS
![image](https://user-images.githubusercontent.com/96085911/153473981-af0afde6-aef9-42d6-9b39-9d682c2f346f.png)

5. Open the file, sort by id
6. There are the duplicates
![image](https://user-images.githubusercontent.com/96085911/153484661-48cb8daf-01e3-46a8-95f5-8d83a651f3cf.png)

7. Open the contact list again
8. Order by id
9. Export to CVS
10. Open the file, sort by id
11. There are no duplicates

#### Mock data to import:
[MOCK_DATA.csv](https://github.com/mautic/mautic/files/8055421/MOCK_DATA.csv)

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Import the mock data (10,000 or more). It can be the data above
3. Open the contact list
4. Leave the default sorting or order by the field with no value (e.g. last_activity if the data was imported)
5. Export to CVS
6. Open file, sort by id
7. There are no duplicates




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10849"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
